parquet-hadoop/src/main/java/org/apache/parquet/hadoop/util/ColumnEncryptor.java

```diff
@@ -0,0 +1,307 @@
+   * @param fileEncryptionProperties FileEncryptionProperties of the file
+   * @throws IOException
+   */
+  public void encryptColumns(String inputFile, String outputFile, List<String> paths, FileEncryptionProperties fileEncryptionProperties) throws IOException {
+    Path inPath = new Path(inputFile);
+    Path outPath = new Path(outputFile);
+
+    ParquetMetadata metaData = ParquetFileReader.readFooter(conf, inPath, NO_FILTER);
+    MessageType schema = metaData.getFileMetaData().getSchema();
+
+    ParquetFileWriter writer = new ParquetFileWriter(HadoopOutputFile.fromPath(outPath, conf), schema, ParquetFileWriter.Mode.OVERWRITE,
+      DEFAULT_BLOCK_SIZE, MAX_PADDING_SIZE_DEFAULT, DEFAULT_COLUMN_INDEX_TRUNCATE_LENGTH, DEFAULT_STATISTICS_TRUNCATE_LENGTH,
+      ParquetProperties.DEFAULT_PAGE_WRITE_CHECKSUM_ENABLED, fileEncryptionProperties);
+    writer.start();
+
+    try (TransParquetFileReader reader = new TransParquetFileReader(HadoopInputFile.fromPath(inPath, conf), HadoopReadOptions.builder(conf).build())) {
+      processBlocks(reader, writer, metaData, schema, paths);
+    }
+    writer.end(metaData.getFileMetaData().getKeyValueMetaData());
+  }
+
+  private void processBlocks(TransParquetFileReader reader, ParquetFileWriter writer, ParquetMetadata meta,
+                            MessageType schema, List<String> encryptPaths) throws IOException {
+    Set<ColumnPath> encryptColumnsPath = convertToColumnPaths(encryptPaths);
+    int blockId = 0;
+    PageReadStore store = reader.readNextRowGroup();
+
+    while (store != null) {
+      writer.startBlock(store.getRowCount());
+
+      List<ColumnChunkMetaData> columnsInOrder = meta.getBlocks().get(blockId).getColumns();
+      Map<ColumnPath, ColumnDescriptor> descriptorsMap = schema.getColumns().stream().collect(
+        Collectors.toMap(x -> ColumnPath.get(x.getPath()), x -> x));
+
+      for (int i = 0; i < columnsInOrder.size(); i += 1) {
+        ColumnChunkMetaData chunk = columnsInOrder.get(i);
+        ColumnDescriptor descriptor = descriptorsMap.get(chunk.getPath());
+        processChunk(descriptor, chunk, reader, writer, encryptColumnsPath, blockId, i, meta.getFileMetaData().getCreatedBy());
+      }
+
+      writer.endBlock();
+      store = reader.readNextRowGroup();
+      blockId++;
+    }
+  }
+
+  private void processChunk(ColumnDescriptor descriptor, ColumnChunkMetaData chunk, TransParquetFileReader reader, ParquetFileWriter writer,
+                            Set<ColumnPath> encryptPaths, int blockId, int columnId, String createdBy) throws IOException {
+    reader.setStreamPosition(chunk.getStartingPos());
+    // If current column chunk is asked to encrypt call encryptPages(), otherwise jut append this column chunk.
+    if (encryptPaths.contains(chunk.getPath())) {
+      writer.startColumn(descriptor, chunk.getValueCount(), chunk.getCodec());
+      encryptPages(reader, chunk, writer, createdBy, blockId, columnId);
+      writer.endColumn();
+    } else {
```

Comment:
@shangxinli per the discussion yesterday, just a thought. What if instead of this `else` block, the same sequence of
```  
      writer.startColumn(descriptor, chunk.getValueCount(), chunk.getCodec());
      encryptPages(reader, chunk, writer, createdBy, blockId, columnId);
      writer.endColumn();
```
is called for unencrypted columns? Of course, `encryptPages` would need to be renamed to something like `copyAndEncryptPages`, and should skip encryption for column that should not be encrypted. But this approach might be able to automatically re-create the column indexes for unencrypted columns, instead of reading and shifting them (in case the latter is challenging).

Reply:
I like the idea to streamline the code path. Only thing is some downgrade of the performance. I see a somewhat 10~20% downgrade comparing with calling appendColumnChunk directly. But I think comparing the risk,  it is worthy of doing that.